core/src/com/unciv/logic/civilization/QuestManager.kt

```diff
@@ -0,0 +1,378 @@
+                else
+                    INDIVIDUAL_QUEST_MIN_TURNS_BETWEEN + Random.nextInt(INDIVIDUAL_QUEST_RAND_TURNS_BETWEEN)
+
+        individualQuestCountdown[challenger.civName] = (countdown * civInfo.gameInfo.gameParameters.gameSpeed.modifier).toInt()
+    }
+
+    private fun tryStartNewGlobalQuest() {
+        if (globalQuestCountdown != 0)
+            return
+        if (assignedQuests.count { it.isGlobal() } >= GLOBAL_QUEST_MAX_ACTIVE)
+            return
+
+        val globalQuests = civInfo.gameInfo.ruleSet.quests.values.filter { it.isGlobal() }
+        val majorCivs = civInfo.getKnownCivs().filter { it.isMajorCiv() && !it.isAtWarWith(civInfo) }
+
+        val assignableQuests = ArrayList<Quest>()
+        for (quest in globalQuests) {
+            val numberValidMajorCivs = majorCivs.count { civ -> isQuestValid(quest, civ) }
+            if (numberValidMajorCivs >= quest.minimumCivs)
+                assignableQuests.add(quest)
+        }
+
+        //TODO: quest probabilities should change based on City State personality and traits
+        if (assignableQuests.isNotEmpty()) {
+            val quest = assignableQuests.random()
+            val assignees = civInfo.gameInfo.getAliveMajorCivs().filter { !it.isAtWarWith(civInfo) && isQuestValid(quest, it) }
+
+            assignNewQuest(quest, assignees)
+            globalQuestCountdown = UNSET
+        }
+    }
+
+    private fun tryStartNewIndividualQuests() {
+        for ((challengerName, countdown) in individualQuestCountdown) {
+            val challenger = civInfo.gameInfo.getCivilization(challengerName)
+
+            if (countdown != 0)
+                return
+
+            if (assignedQuests.count { it.assignee == challenger.civName && it.isIndividual() } >= INDIVIDUAL_QUEST_MAX_ACTIVE)
+                return
+
+            val assignableQuests = civInfo.gameInfo.ruleSet.quests.values.filter { it.isIndividual() && isQuestValid(it, challenger) }
+            //TODO: quest probabilities should change based on City State personality and traits
+            if (assignableQuests.isNotEmpty()) {
+
+                val quest = assignableQuests.random()
+                val assignees = arrayListOf(challenger)
+
+                assignNewQuest(quest, assignees)
+            }
+        }
+    }
+
+    private fun handleGlobalQuests() {
+        val globalQuestsExpired = assignedQuests.filter { it.isGlobal() && it.isExpired() }.map { it.questName }.distinct()
+        for (globalQuestName in globalQuestsExpired)
+            handleGlobalQuest(globalQuestName)
+    }
+
+    private fun handleGlobalQuest(questName: String) {
+        val quests = assignedQuests.filter { it.questName == questName }
+        if (quests.isEmpty())
+            return
+
+        val topScore = quests.map { getScoreForQuest(it) }.max()!!
+
+        for (quest in quests) {
+            if (getScoreForQuest(quest) >= topScore)
+                giveReward(quest)
+        }
```

Comment:
Here we call getScoreForQuest twice for each quest - once to get the max, and a second time to check if it's the max.
A more concise way would be 
> for(quest in quests.groupBy { getScoreForQuest(it) }.entries.maxBy { it.key }!!.value) giveReward(quest)
But I'm not sure if that's readable to the average contributor...

Reply:
Yup, i preferred readability over optimization here. As you can see, getScoreForQuest is just a subtraction: it shouldn't impact performance too much. If you prefer the way you suggested let me know